service/topology_coordinator.cc

```diff
@@ -2066,6 +2079,137 @@ future<bool> topology_coordinator::maybe_start_tablet_migration(group0_guard gua
     co_return true;
 }
 
+future<bool> topology_coordinator::do_upgrade_step(group0_guard guard) {
+    switch (_topo_sm._topology.ustate) {
+    case topology::upgrade_state::not_upgraded:
+        on_internal_error(rtlogger, std::make_exception_ptr(std::runtime_error(
+                "topology_coordinator was started even though upgrade to raft topology was not requested yet")));
+
+    case topology::upgrade_state::build_coordinator_state:
+        co_await build_coordinator_state(std::move(guard));
+        co_return true;
+
+    case topology::upgrade_state::final_global_barrier: {
+        rtlogger.info("upgrade complete, performing a global barrier before continuing");
+        guard = co_await exec_global_command(std::move(guard),
+                raft_topology_cmd{raft_topology_cmd::command::barrier},
+                {_raft.id()});
+
+        topology_mutation_builder builder(guard.write_timestamp());
+        builder.set_upgrade_state(topology::upgrade_state::done);
+        co_await update_topology_state(std::move(guard), {builder.build()}, "upgrade: marking as complete");
+
+        co_return true;
+    }
+
+    case topology::upgrade_state::done:
+        on_internal_error(rtlogger, std::make_exception_ptr(std::runtime_error(
+                "topology_coordinator::do_upgrade_step called after upgrade was completed")));
+    }
+}
+
+future<> topology_coordinator::build_coordinator_state(group0_guard guard) {
+    // Wait until all nodes reach use_post_raft_procedures
+    rtlogger.info("waiting for all nodes to finish upgrade to raft schema");
+    co_await _group0.wait_for_all_nodes_to_finish_upgrade(_as);
+
+    // Execute a global barrier to make sure that all nodes see
+    // the build_coordinator_state upgrade_state
```

Comment:
Why all nodes have to see the state before we proceed with the upgrade?

Reply:
After the barrier, the whole cluster will be put into a state where no legacy topology operations are allowed anymore. This prevents a situation where some nodes allow initiating legacy topology operations and some others already operate in raft mode.

To be fair... Given that we already require administrators to abstain from issuing topology requests during the procedure - and they need to check that all nodes ultimately see the "done" state - this doesn't seem to bring much value, so maybe this precaution can be dropped. On the other hand, it doesn't cost us much to do it (no additional states, RPC verbs etc.). @kbr-scylla WDYT?

Reply:
> To be fair... Given that we already require administrators to abstain from issuing topology requests during the procedure - and they need to check that all nodes ultimately see the "done" state - this doesn't seem to bring much value

Exactly.

Reply:
Such precautions don't hurt, and they will be needed once we make the procedure fully automatic and safe from administrator errors (that is the plan, remember?) so why not add it now

Reply:
Because we cannot yet know where they are needed and for what. It may be not enough or not needed here.

Reply:
We do know because we started by designing the full procedure

@piodul isn't this one obvious? We know we'll need it. We want all nodes to learn that we're in building phase to disable topology/schema changes

Reply:
We don't need it right now, and adding it later shouldn't be a problem.

Right now it makes people trying to understand the current procedure, but who are not familiar with the full procedure, scratch their head.

Reply:
Lets add things when they are needed. This way we will know why  they are needed.